pandas/_typing.py

```diff
@@ -222,7 +222,7 @@ def flush(self) -> Any:
 
 
 class ReadPickleBuffer(ReadBuffer[bytes], Protocol):
-    def readline(self) -> AnyStr_cov:
+    def readline(self) -> bytes:
```

Comment:
I think `readline(self)` needs to be `readline(self, size: int | None = ...) -> bytes`

https://docs.python.org/3/library/io.html#io.IOBase.readline

Reply:
typeshed has both, with and without an optional argument: https://github.com/python/typeshed/blob/9cda16f5df8386c8cf9542dda2d3cb83abf4fd55/stdlib/_typeshed/__init__.pyi#L212

If we don't require the argument, we still accept taking objects that can be called without providing an argument. But if we require an argument, we do not accept objects than cannot take an argument. 

- https://github.com/python/typeshed/blob/1dc1c91429b921e6444c677b9197d5b3f5b0f064/stdlib/io.pyi#L146 accepts an argument but it cannot be None
- https://github.com/python/typeshed/blob/de1a79bd007e73bd57821a89e21bc08d4d07e75a/stdlib/imaplib.pyi#L55 accepts no argument
- https://github.com/python/typeshed/blob/ceccc655dbeaaf3540fc2f34fc377f974e3447e9/stdlib/fileinput.pyi#L302 accepts no argument

Reply:
The following works without needing to take an optional argument:

```py
from io import BytesIO
import numpy as np
import pandas as pd

class Reader:
    def __init__(self, content: bytes) -> None:
        self.buffer = BytesIO(content)

    @property
    def mode(self) -> str:
        return "rb"

    def read(self, n: int=-1) -> bytes:
        return self.buffer.read(n)

    def seek(self, pos: int) -> None:
        return self.buffer.seek(pos)

    def readline(self) -> None:
        return self.buffer.readline()

df = pd.DataFrame(np.random.rand(100, 10))
buffer = BytesIO()
df.to_pickle(buffer)
print(pd.read_pickle(Reader(buffer.getvalue())))
```

Reply:
When I tried

```
with open("file.pkl", "rb") as f:
    pd.read_pickle(f)
```

it complained about the second argument.

Reply:
On main: mypy will print the offending protocol methods, but the offending part was the return type, not the second argument (even though both look different).

The following works with this PR, with both mypy and pyright:

```py
# pyright: strict
import tempfile

from pandas._typing import ReadPickleBuffer

import pandas as pd


def read(x: ReadPickleBuffer) -> None:
    ...


with tempfile.NamedTemporaryFile(delete=False) as file:
    read(file)
    pd.read_pickle(file)

with open("file.pkl", mode="rb") as f:
    read(f)
    pd.read_pickle(f)
```

Reply:
@bashtage were you able to test it with this PR?

Reply:
Pandas-stubs has since https://github.com/pandas-dev/pandas-stubs/pull/326 the same protoclls as in this PR: its tests still pass and so far there hasn't been an issue raised that it broke something.